src/controls/controls.js

```diff
@@ -155,15 +162,35 @@ Crafty.c("Multiway", {
             .unbind("KeyUp", this._keyup);
 
         // unapply movement of pressed keys
-        this.__unapplyActiveDirections();
+        this.__unapplyActiveSpeed();
     },
 
     _keydown: function (e) {
         var direction = this._keyDirection[e.key];
         if (direction !== undefined) { // if this is a key we are interested in
-            if (this._activeDirections[direction] === 0 && !this.disableControls) { // if key is first one pressed for this direction
-                this.vx += this._directionSpeed[direction].x;
-                this.vy += this._directionSpeed[direction].y;
+            if (this._activeDirections[direction] === 0) { // if key is first one pressed for this direction
+                /* Following is an optimization of these statements:
+                if (!this.disableControls) this.__unapplyActiveSpeed();
+                this.__addActiveMagnitude(direction);
+                this.__updateActiveSpeed();
+                if (!this.disableControls) this.__applyActiveSpeed();
+                */
+                var oldActiveSpeedX = this._activeSpeedX,
+                    oldActiveSpeedY = this._activeSpeedY;
+
+                var magnitude_e6 = this._directionMagnitudes_e6[direction],
```

Comment:
Each direction vector is added to the net movement vector when key is pressed.
Each direction vector is substracted from the net movement vector when key is released.
Pressing and then releasing 5 different direction keys should net into 0 movement.
However, due to rounding errors, you don't get 0.
That's why this net movement vector is represented as a number `* 10e6`.